### ISSUE TITLE
Serialize to and deserialize from starfile

### ISSUE THREAD

**Issue by mjo22** [mjo22/cryojax#8]
`ImagePipeline`s and other cryojax objects should serialize to and deserialize from standard cryo-EM file formats, such as Relion starfiles.

**Comment by DSilva27** [comment]
Implemented a dataloader using starfile, based on the dataloader used in [aNiMAtE](https://github.com/compSPI/aNiMAte/blob/main/src/experiment_scripts/generate_dataset.py) and an adaptation of PyTorch's dataloader to work with numpy (see [https://jax.readthedocs.io/en/latest/notebooks/Neural_Network_and_Data_Loading.html](https://jax.readthedocs.io/en/latest/notebooks/Neural_Network_and_Data_Loading.html))

**Comment by DSilva27** [comment]
Opened pull request for this issue #50

**Comment by DSilva27** [comment]
Need to fix the "super" in the data loader classes.

**Comment by mjo22** [comment]
This looks awesome. At some point, maybe it would be really cool to have the ImagePipeline (or maybe the Pose? the CTF?) have methods to_starfile and from_starfile?

**Comment by DSilva27** [comment]
Yes, I will work on implementing a `to_starfile`. If you want I can also implement a function that takes a starfile filename, and a set of keys (pose, shifts, etc) and retrieves the desired values from the starfile.

**Comment by mjo22** [comment]
This would be awesome. It sounds like this function could be used in a `from_starfile` routine to directly instantiate `Pose`s and `Optics` models? Once we have this I can think about how/where to add these methods…

**Comment by mjo22** [comment]
On this same coin, maybe it would be good to have some sort of `from_starfile_loader` (not sure about the name) function that loads `Pose` and `Optics` models, whose parameters have a leading axis representing the batch dimension. These can then be used directly with vmap (see here https://docs.kidger.site/equinox/tricks/#ensembling)

**Comment by mjo22** [comment]
See the starfile branch for this issue. Also, ignore my above comments. I now disagree with this… I’m taking the philosophy that all bound methods of a class should work with JAX transformations

**Comment by mjo22** [comment]
Okay, I am getting around to this finally. I have been reading through the PR and think I have a variant of implementing the dataloader that will be maximally flexible and easy to install. This will entail something like the following schematic

```python
import equinox as eqx
import dataclasses
import starfile
from jaxtyping import Shaped
from cryojax.simulator import EulerAnglePose, CTF, ImageConfig
from typing import RealImage

@dataclasses.dataclass(frozen=True)
class RelionDataset:

    def __init__(path_to_starfile):
        # run starfile.read to load starfile

    def __getattr___(self, idx) -> ParticleStack:
        # create a ParticleStack pytree that can be
        # passed to jax transformations

    def __len__(self) -> int:
        # get length

class ParticleStack(eqx.Module):

    image_stack: Shaped[RealImage, "batch_dim"]
    config: Shaped[ImageConfig, "batch_dim"]
    pose: Shaped[EulerAnglePose, "batch_dim"]
    ctf: Shaped[CTF, "batch_dim"]
```

Now, one wants to use a `torch` dataloader, one would externally create a `torch.utils.data.Dataset` by wrapping the above `RelionDataset`

```python
from torch.utils.data import Dataset

class TorchRelionDataset(Dataset):

    def __init__(cryojax_dataset: RelionDataset):
        self.cryojax_dataset = cryojax_dataset

    def __getattr___(self, idx) -> ParticleStack:
        return self.cryojax_dataset[idx]

    def __len__(self) -> int:
        return len(self.cryojax_dataset)

```

This way, cryojax need not assume a particular pattern for how to implement the dataloader. Thoughts?

**Comment by mjo22** [comment]
Made an inital commit for a `cryojax.data` submodule in the starfile branch. The plan is to take chunks of code from the dataloader PR (#50), rather than accepting it directly. The reason for this is the larger restructuring i am making with `cryojax.data`.
